File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/autos/AutoHardware.java
package org.firstinspires.ftc.teamcode.autos;

import com.qualcomm.robotcore.eventloop.opmode.LinearOpMode;
import com.qualcomm.robotcore.hardware.DcMotor;
import com.qualcomm.robotcore.util.Range;

import org.firstinspires.ftc.teamcode.Hardware;

public class AutoHardware extends Hardware {
    public AutoHardware(LinearOpMode opMode, boolean camera) throws InterruptedException {
        super(opMode, camera);
        initWheels();
    }

    // Driving -------------------------------------------------------------------------------------

    private void initWheels() {
        leftFront.setTargetPositionTolerance(15);
        leftBack.setTargetPositionTolerance(15);
        rightFront.setTargetPositionTolerance(15);
        rightBack.setTargetPositionTolerance(15);
    }

    public void setVelocity(double speed) {
        setWheelPower(1);
        setAllMotorModes(DcMotor.RunMode.RUN_USING_ENCODER);
        leftFront.setVelocity(speed);
        leftBack.setVelocity(speed);
        rightFront.setVelocity(speed);
        rightBack.setVelocity(speed);
    }

    private static final double TICKS_PER_ROT = 537.6;
    private static final double WHEEL_DIAMETER = 5;
    private static final double INCHES_PER_ROT = 3.14159 * WHEEL_DIAMETER;
    private static final double TICKS_PER_INCH = TICKS_PER_ROT / INCHES_PER_ROT;

    public void driveInches(double inches) throws InterruptedException {
        setWheelPower(1);
        setAllMotorModes(DcMotor.RunMode.STOP_AND_RESET_ENCODER);
        setAllTargetPosition(inchesToTicks(inches));
        setAllMotorModes(DcMotor.RunMode.RUN_TO_POSITION);
        waitForEncoders();
    }

    // -----

//    private static final double TICKS_PER_DEGREE = 5.85;
    private static final double TICKS_PER_DEGREE = 5.27;

    public void turnDegrees(double degrees) throws InterruptedException {
        setWheelPower(1);
        setAllMotorModes(DcMotor.RunMode.STOP_AND_RESET_ENCODER);
        int ticks = (int) (degrees * TICKS_PER_DEGREE);
        setTargetPositions(ticks, -ticks);
        setAllMotorModes(DcMotor.RunMode.RUN_TO_POSITION);
        waitForEncoders();
    }

    public void goToHeading(double angle) throws InterruptedException {
        setAllMotorModes(DcMotor.RunMode.STOP_AND_RESET_ENCODER);
        setTargetPositionBasedOnHeading(angle);
        setAllMotorModes(DcMotor.RunMode.RUN_TO_POSITION);
        int correctCount = 0;
        while (!opMode.isStopRequested() && correctCount < 20) {
            double error = headingError(angle);
            setWheelPower(Range.clip(Math.abs(error) / 180.0, 0.08, 1.0) * 0.6);
            setTargetPositionBasedOnHeading(angle);
//            Thread.sleep(10);
            if (Math.abs(headingError(angle)) < 1) {
                correctCount++;
            } else {
                correctCount = 0;
            }
        }
    }

    private void setTargetPositionBasedOnHeading(double angle) {
        double error = headingError(angle);
        if (error < 0) error += 360;
        if (error > 180) error -= 360;
        int ticks = (int) (error * TICKS_PER_DEGREE);
//        setTargetPositions(ticks, -ticks);
        leftFront.setTargetPosition(leftFront.getCurrentPosition() + ticks);
        leftBack.setTargetPosition(leftBack.getCurrentPosition() + ticks);
        rightFront.setTargetPosition(rightFront.getCurrentPosition() - ticks);
        rightBack.setTargetPosition(rightBack.getCurrentPosition() - ticks);
        opMode.telemetry.addData("ticks", ticks);
        opMode.telemetry.addData("heading", robotHeading());
        opMode.telemetry.addData("error", error);
        opMode.telemetry.addData("lf t", leftFront.getTargetPosition());
        opMode.telemetry.addData("lf a", leftFront.getCurrentPosition());
        opMode.telemetry.update();

    }

    public double headingError(double targetAngle) {
        return (targetAngle - robotHeading()) % 360;
    }

    // -----

    public boolean isEncoderDriving() {
        return leftBack.isBusy() || leftFront.isBusy() || rightBack.isBusy() || rightFront.isBusy();
    }

    public void waitForEncoders() throws InterruptedException {
        while (isEncoderDriving()) {
            displayWheelTelemetry();
            Thread.sleep(100);
        }
    }

    private int inchesToTicks(double inches) {
        return (int) (inches * TICKS_PER_INCH);
    }

    private void setTargetPositions(int lTicks, int rTicks) {
        leftFront.setTargetPosition(lTicks);
        leftBack.setTargetPosition(lTicks);
        rightFront.setTargetPosition(rTicks);
        rightBack.setTargetPosition(rTicks);
    }

    private void setAllTargetPosition(int ticks) {
        setTargetPositions(ticks, ticks);
    }

    private void setAllMotorModes(DcMotor.RunMode mode) {
        leftFront.setMode(mode);
        leftBack.setMode(mode);
        rightFront.setMode(mode);
        rightBack.setMode(mode);
    }

    public void displayWheelTelemetry() {
        opMode.telemetry.addData("left front", leftFront.getTargetPosition() - leftFront.getCurrentPosition());
        opMode.telemetry.addData("left back", leftBack.getTargetPosition() - leftBack.getCurrentPosition());
        opMode.telemetry.addData("right front", rightFront.getTargetPosition() - rightFront.getCurrentPosition());
        opMode.telemetry.addData("right back", rightBack.getTargetPosition() - rightBack.getCurrentPosition());
        opMode.telemetry.update();
    }

    // Grabber -------------------------------------------------------------------------------------

    @Override
    public void setGrabberIsOpen(boolean isOpen) throws InterruptedException {
        super.setGrabberIsOpen(isOpen);
        Thread.sleep(300);
    }

    @Override
    protected void setGrabberTargetPosition() throws InterruptedException {
        super.setGrabberTargetPosition();
        while (grabberMotor.isBusy()) {
            Thread.sleep(100);
        }
    }

}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/autos/AutoHardware.kt
package org.firstinspires.ftc.teamcode.autos

import com.qualcomm.robotcore.eventloop.opmode.LinearOpMode
import com.qualcomm.robotcore.hardware.DcMotor
import com.qualcomm.robotcore.hardware.DcMotor.RunMode
import com.qualcomm.robotcore.hardware.DistanceSensor
import com.qualcomm.robotcore.util.ElapsedTime
import org.firstinspires.ftc.robotcore.external.navigation.DistanceUnit
import org.firstinspires.ftc.teamcode.utils.Hardware
import kotlin.math.abs
import kotlin.math.max

class AutoHardware(opMode: LinearOpMode, camera: Boolean) : Hardware(opMode, camera) {

    init {
        initWheels()
    }

    fun frontDistance(): Double = frontDistanceSensor.getDistance(DistanceUnit.INCH)
    fun backDistance(): Double = backDistanceSensor.getDistance(DistanceUnit.INCH)
    fun frontColorDistance(): Double = colorDist.getDistance(DistanceUnit.INCH)

    private val frontDistanceSensor = opMode.hardwareMap[DistanceSensor::class.java, "dist front"]
    private val backDistanceSensor = opMode.hardwareMap[DistanceSensor::class.java, "dist back"]
    private val colorDist = opMode.hardwareMap[DistanceSensor::class.java, "front color"]

    // Driving -------------------------------------------------------------------------------------
    private fun initWheels() {
        leftFront.targetPositionTolerance = 15
        leftBack.targetPositionTolerance = 15
        rightFront.targetPositionTolerance = 15
        rightBack.targetPositionTolerance = 15

        leftFront.zeroPowerBehavior = DcMotor.ZeroPowerBehavior.BRAKE
        leftBack.zeroPowerBehavior = DcMotor.ZeroPowerBehavior.BRAKE
        rightFront.zeroPowerBehavior = DcMotor.ZeroPowerBehavior.BRAKE
        rightBack.zeroPowerBehavior = DcMotor.ZeroPowerBehavior.BRAKE
    }

    fun setVelocity(speed: Double) {
        setWheelPower(1.0)
        setAllMotorModes(RunMode.RUN_USING_ENCODER)
        leftFront.velocity = speed
        leftBack.velocity = speed
        rightFront.velocity = speed
        rightBack.velocity = speed
    }

    fun driveInches(inches: Double, speedIn: Double = 0.8) {
        val speed = abs(speedIn).coerceIn(0.0, 1.0)
        setWheelPower(speed)
        setAllMotorModes(RunMode.STOP_AND_RESET_ENCODER)
        setAllTargetPosition(inchesToTicks(inches))
        setAllMotorModes(RunMode.RUN_TO_POSITION)

        val startAngle = robotHeading()
        while (!opMode.isStopRequested && anyMotorBusy) {
            // adjust relative speed based on heading error.
            val error = headingError(startAngle)
            var steer = (error * DRIVE_P).coerceIn(-1.0, 1.0)
//            var steer = 0.0

            // if driving in reverse, the motor correction also needs to be reversed
            if (inches < 0) steer *= -1.0

            var leftSpeed = (speed + steer).addToAbs(MIN_DRIVE_SPEED)
            var rightSpeed = (speed - steer).addToAbs(MIN_DRIVE_SPEED)

            // Normalize speeds if either one exceeds +/- 1.0;
            val max = max(abs(leftSpeed), abs(rightSpeed))
            if (max > 1.0) {
                leftSpeed /= max
                rightSpeed /= max
            }

            setWheelPower(leftSpeed, rightSpeed)
        }

        setAllMotorModes(RunMode.RUN_WITHOUT_ENCODER)
        setWheelPower(0.0)
        turnToHeading(startAngle)
        Thread.sleep(500)
    }


    private fun Double.addToAbs(min: Double, max: Double = 1.0) = when {
        this >= 0 -> (this + min).coerceAtMost(max)
        else -> (this - min).coerceAtLeast(-max)
    }

    // TURNING: ------------------------------------------------------------------------------------

    fun turnToMinDist(minDist: Double, direction: Double, speed: Double = 0.3) {
        setWheelPower(speed * direction, -speed * direction)

        println("----------------")
        var dist = frontDistance()
        println(dist)
        while (opMode.opModeIsActive() && dist > minDist) {
            println(dist)
            dist = frontDistance()
        }
        println(">>$dist")
        setWheelPower(0.0)
    }

    fun turnToHeading(angle: Double, log: Boolean = false) {
        val time = ElapsedTime()
        setAllMotorModes(RunMode.RUN_WITHOUT_ENCODER)

        var error = headingError(angle)

        val totalTime = ElapsedTime()
        val loopTime = ElapsedTime()

        var lastError = 0.0
        var integralSum = 0.0
        var loops = 0

        while ((time.milliseconds() < 1000) && !opMode.isStopRequested && (abs(error) > TURN_TOLERANCE || !wheelsAreStopped())) {
            error = headingError(angle)
            if (error > 5 || error < -5) time.reset()
            val derivative = (error - lastError) / loopTime.seconds()
            integralSum += error * loopTime.seconds()

            opMode.telemetry.addData("loops", loops++)
            opMode.telemetry.addData("error", error)
            opMode.telemetry.addData("time", time.milliseconds())
            opMode.telemetry.update()
            if (log) println("${totalTime.seconds()}, $error")

            val power = (TURN_P * error + TURN_I * integralSum + TURN_D * derivative).addToAbs(
                MIN_TURN_SPEED, MAX_TURN_SPEED
            ) * 0.8
            setWheelPower(power, -power)

            lastError = error
            loopTime.reset()
        }
        if(log) println("STOPPED")
        setWheelPower(0.0)
    }

    private fun wheelsAreStopped(): Boolean = true
//        leftFront.velocity < 50.0 &&
//                leftBack.velocity < 50.0 &&
//                rightFront.velocity < 50.0 &&
//                rightBack.velocity < 50.0

    private fun headingError(toAngle: Double): Double {
        return normalizeHeading(toAngle - robotHeading())
    }

    private fun normalizeHeading(angle: Double): Double {
        var correctedAngle = angle
        while (correctedAngle > 180) correctedAngle -= 360.0
        while (correctedAngle < -180) correctedAngle += 360.0
        return correctedAngle
    }

    // ---------------------------------------------------------------------------------------------

    val isEncoderDriving: Boolean
        get() = leftBack.isBusy || leftFront.isBusy || rightBack.isBusy || rightFront.isBusy

    val anyMotorBusy: Boolean
        get() = leftBack.isBusy && leftFront.isBusy && rightBack.isBusy && rightFront.isBusy

    private fun inchesToTicks(inches: Double): Int {
        return (inches * TICKS_PER_INCH).toInt()
    }

    private fun setTargetPositions(lTicks: Int, rTicks: Int) {
        leftFront.targetPosition = lTicks
        leftBack.targetPosition = lTicks
        rightFront.targetPosition = rTicks
        rightBack.targetPosition = rTicks
    }

    private fun setAllTargetPosition(ticks: Int) {
        setTargetPositions(ticks, ticks)
    }

    private fun setAllMotorModes(mode: RunMode) {
        leftFront.mode = mode
        leftBack.mode = mode
        rightFront.mode = mode
        rightBack.mode = mode
    }

    // Grabber -------------------------------------------------------------------------------------

    override fun setGrabberIsOpen(isOpen: Boolean) {
        super.setGrabberIsOpen(isOpen)
        Thread.sleep(700)
    }

    override fun stowGrabbers() {
        super.stowGrabbers()
        Thread.sleep(700)
    }

    override fun setGrabberTargetPosition() {
        super.setGrabberTargetPosition()
        while (grabberMotor.isBusy) {
            Thread.sleep(100)
        }
    }

    companion object {
        private const val TICKS_PER_ROT = 537.6
        private const val WHEEL_DIAMETER = 5.0
        private const val INCHES_PER_ROT = 3.14159 * WHEEL_DIAMETER
        private const val TICKS_PER_INCH = TICKS_PER_ROT / INCHES_PER_ROT

        private const val DRIVE_P = 0.07
        private const val MIN_DRIVE_SPEED = 0.3

        private const val TURN_P = 0.04

        //        private const val TURN_I = 0.0004
//        private const val TURN_D = 0.0025
        private const val TURN_I = 0
        private const val TURN_D = 0.001
        private const val MIN_TURN_SPEED = 0.2
        private const val MAX_TURN_SPEED = 0.5
        private const val TURN_TOLERANCE = 1.0

    }
}